final-project/application.py

```diff
@@ -486,6 +486,84 @@ def search_category():
         users_selected_category=users_selected_category)
 
 
+# Store references to product research
+@app.route("/reference")
+def reference():
+    """Search for references to product research."""
+
+    # Keep the user logged in
+    user_id = session.get("user_id")
+
+    # Get the product name requested by the users
+    product_to_provide_references_for = request.args.get(
+            "product")
+
+    # If no name is given, then ask the user for one.
+    if not product_to_provide_references_for:
+        return apology("Please specify product")
+    elif product_to_provide_references_for == None:
+        return apology("Please specify product")
+
+    # Search for references to a given product
+    references_for_product = db.execute("""
+            SELECT research.link, title
+            FROM research
+            LEFT JOIN products
+            ON products.id=research.product_id
+            WHERE products.product_name=:product
+            """, product=product_to_provide_references_for)
+
+    # Check whether we found references
+    if not references_for_product:
+        return apology("We have no references. Please help us research")
+    elif references_for_product == None:
+        return apology("We have no references. Please help us research")
+
+    # Save the number of research articles for this products.
+    number_of_references = len(references_for_product)
+
+    # Save the titles and links to each reference.
+    # Create an empty list to save our titles.
+    reference_titles = []
+    # Create an empty list to save our links.
+    reference_links = []
+
+    # Loop through each title in our list of references
+    for reference in range(number_of_references):
+
+        # Ensure we have a title.
+        if not references_for_product[reference]["title"]:
+            # If there is no title, then use the link.
+            reference_titles.append(references_for_product[reference]["link"])
+        # If there is a title, then use the title.
+        else:
+            # Save the title.
+            reference_titles.append(references_for_product[reference]["title"])
+
+        # Save the link.
+        reference_links.append(references_for_product[reference]["link"])
+
+
+    # Send product research references to display for the user.
+    return render_template("references.html"
+            , number_of_references=number_of_references
+            , reference_titles=reference_titles
+            , reference_links=reference_links
+            , product_to_provide_references_for=product_to_provide_references_for)
+
+    return apology("TODO")
+
+
+# TODO: Add references to product research
+@app.route("/add_reference")
```

Comment:
Have you learned about the REST API format? I notice a lot of your routes are named liked methods. This isn't the mainstream convention, and I'd take some time to learn the standard.

I'm sure there's a good resource somewhere here: https://www.youtube.com/results?search_query=rest+api&page=&utm_source=opensearch

Reply:
I remember seeing REST appear like wildfire on Hacker News a year back, but that is the extent of it. I was always curious. I will look more into it.